design/one-pager-package-image-config.md

```diff
@@ -0,0 +1,377 @@
+its `ClusterImagePolicy` API, Policy Controller allows users to define policies
+for verifying images with a wide range of options. As of today, Policy
+Controller only operates on Pods or the native Kubernetes resources resulting in
+a Pod, e.g. Deployments, StatefulSets, etc. There is no way to configure it to
+work with Crossplane packages.
+
+Kyverno, on the other hand, is a general-purpose policy engine for Kubernetes
+that allows users to define policies for any Kubernetes resource. Using its
+`ClusterPolicy` API, one can define [rules to verify images] and by leveraging
+the `imageExtractors` field, it is possible to configure it to work with
+Crossplane packages. See the [using Kyverno for package image verification]
+section for more details.
+
+Another example in this space is the project [Flux]. Compared to the previous
+APIs, this project provides a [minimalistic API to verify images] with the
+`OCIRepository` API, marking some parts of the API (e.g. keyless verification)
+as experimental.
+
+In Crossplane itself, we had [a previous proposal] aiming to solve signature
+verification for packages but mostly focusing on the implementation details,
+primarily for verification using a shared public key. The follow-up
+[implementation PR] extended the solution also to support keyless verification
+by mostly following the same solution / similar API as Flux by extending the
+Package installation APIs for verification settings, leaving the dependencies
+out of the scope. This proposal aims to provide a more comprehensive solution by
+introducing an API that can be extended to support various verification settings
+and also covering the ones coming as a dependency.
+
+## Goals
+
+- Allow configuration of credentials for package images and their dependencies
+  at runtime.
+- Enable users to set policies for verifying image signatures.
+- Define an image-centric configuration API that centralizes settings for
+  package images.
+
+## Proposal
+
+This document proposes a new API, `ImageConfig`, under the `pkg.crossplane.io`
+API group, that allows users to configure settings for package images. The API
+enables users to define rules for matching images and configuring how to
+interact with the registries hosting the images, including authentication and
+TLS settings. It also provides a way to define policies for verifying image
+signatures.
+
+Considering the API is solving some existing problems, e.g. configuring pull
+secrets for dependencies, and not introducing a behavioral change unless the
+users opt in by creating an `ImageConfig` object, we believe this API should be
+introduced as a beta API without going through the feature lifecycle stages.
+
+### API
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: ImageConfig
+metadata:
+  name: acme-packages
+spec:
+  matchImages:
+    - prefix: registry1.com/acme-co/configuration-foo
+    - prefix: registry1.com/acme-co/configuration-bar
+    - prefix: registry1.com/acme-co/function-baz
```

Comment:
Please see: https://github.com/crossplane/crossplane/pull/5993#discussion_r1794839623

I don't think we will need to support other options, but I'm cannot be 100% sure, of course. Not regexp/glob, but, may be a constraint could be helpful, which would be a supportive field rather than an alternative.

```
matchImages:
  - prefix: registry1.com/acme-co/configuration-foo
     constraint: "=< v1"
```

Reply:
I'll bet we'll end up with regex support 😆 kyverno does that too [afaict](https://kyverno.io/docs/writing-policies/verify-images/sigstore/#verifying-image-signatures)

Reply:
Container registries only support prefixes, except subdomain wildcard: https://www.mankier.com/5/containers-registries.conf#Description-Choosing_a_%5B%5Bregistry%5D%5D_TOML_table

Reply:
The safest thing to do would be to include a type distinguisher field, e.g.:

```yaml
matchImages:
- match: Prefix
  prefix: registry1.com/acme-co/configuration-foo
```

If there ends up being multiple options, the benefits of a type distinguisher field are:

* It's less surprising - you always explicitly pick a match type writing the config. There's no auto defaulting.
* It's self documenting - you always see a match when reading the config. This helps you discover that there might be other match types for you to consider.

If we never end up with other options, the downside is that everyone has to write `match: Prefix` boilerplate for no real reason.

To me the benefit outweighs the cost. I learned when working on the P&T APIs that I'd typically always regret not having made `type` a required field from the start.

That said, I'll defer to you on the final call on this @turkenh. It sounds like you have high confidence that we'll never need to support anything other than `prefix`.

Reply:
This sounds good to me, I'll add a type distinguisher field.